node/subsystem-util/src/rolling_session_window.rs

```diff
@@ -354,13 +503,23 @@ async fn get_session_index_for_child(
 	}
 }
 
-async fn load_all_sessions(
+/// Attempts to extend db stored sessions with sessions missing between `start` and up to `end_inclusive`.
+/// Runtime session info fetching errors are ignored if that doesn't create a gap in the window.
+async fn extend_sessions_from_chain_state(
+	stored_sessions: Vec<SessionInfo>,
 	sender: &mut impl overseer::SubsystemSender<RuntimeApiMessage>,
 	block_hash: Hash,
```

Comment:
I don't think this is going to work. The runtime of `block_hash` only store the last `dispute_period` session. What we want to query the runtime of different blocks from old session.

Reply:
In usual scenarios however, this function would be only called when a new session happens. So, applying the said fix would make it work when we have a long finality stall across more sessions than `dispute_period` and the node has been offline for the same amount of sessions as well.

Reply:
Many node operators also sync from a snapshot that omits the parachain db. Some wipe the db completely and warp-sync from scratch. Surely, we don't expect that to happen on mainnets en masse, but we better be prepared in case it does. This is especially annoying on small testnets with short session durations.

Reply:
> IIUC we would need to stop and not go beyond the last finalized block, as the runtime information would not be available anymore, right 

Yep, there's no need to go below finalized block, as we won't be voting on anything below including disputes and will prune the state as you mentioned as well as parachain db state.

Reply:
> I've been thinking more about this case If there is no parachains db, we'd also miss the approval votes DB. 

In case there's a dispute, some of the votes will be posted on chain, so they will be recoverable without parachain db.
Empty parachain db was just one example. 
I still think it would be good to handle this (unlikely) edge case properly where some validators are being offline for a few sessions, will open a follow-up issue.